arch/arm/src/imxrt/imxrt_flexio.c

```diff
@@ -0,0 +1,851 @@
+ *
+ * param dev FlexIO device
+ * param mask The timer status mask which can be calculated by
+ *            (1 << timer index)
+ *
+ * note For clearing multiple timer status flags, for example, two timer
+ * status flags, can calculate the mask by using ((1 << timer index0) |
+ * (1 << timer index1))
+ */
+
+static void imxrt_flexio_clear_timer_status_flags(
+  struct flexio_dev_s *dev,
+  uint32_t mask)
+{
+  struct flexio_type_s *base = ((struct imxrt_flexiodev_s *)dev)->base;
+
+  base->TIMSTAT = mask;
+}
+
+/* Enables/disables the shifter status DMA. The DMA request generates when
+ * the corresponding SSF is set.
+ *
+ * For multiple shifter status DMA enables, for example, calculate
+ * the mask by using ((1 << shifter index0) | (1 << shifter index1))
+ *
+ * param dev FlexIO device
+ * param mask The shifter status mask which can be calculated by
+ *            (1 << shifter index)
+ * param enable True to enable, false to disable.
+ */
+
+static void imxrt_flexio_enable_shifter_status_dma(
+  struct flexio_dev_s *dev,
+  uint32_t mask,
+  bool enable)
+{
+  struct flexio_type_s *base = ((struct imxrt_flexiodev_s *)dev)->base;
+
+  if (enable)
+    {
+      base->SHIFTSDEN |= mask;
+    }
+  else
+    {
+      base->SHIFTSDEN &= ~mask;
+    }
+}
+
+/* Configures the FlexIO with a FlexIO configuration. The configuration
+ * structure can be filled by the user or be set with default values by
+ * imxrt_flexio_get_default_config().
+ *
+ * Example
+ * code
+ * struct flexio_config_s config = {
+ * .enable_flexio = true,
+ * .enable_indoze = false,
+ * .enable_indebug = true,
+ * .enable_fast_access = false
+ * };
+ * imxrt_flexio_init(base, &config);
+ * endcode
+ *
+ * param dev FlexIO device
+ * param user_config pointer to flexio_config_s structure
+ */
+
+static void imxrt_flexio_init(
+  struct flexio_dev_s *dev,
+  const struct flexio_config_s *user_config)
+{
+  struct flexio_type_s *base = ((struct imxrt_flexiodev_s *)dev)->base;
+  uint32_t ctrl_reg = 0;
+
+  imxrt_flexio_reset(dev);
+
+  ctrl_reg = base->CTRL;
+  ctrl_reg &= ~(FLEXIO_CTRL_DOZEN_MASK |
+                FLEXIO_CTRL_DBGE_MASK |
+                FLEXIO_CTRL_FASTACC_MASK |
+                FLEXIO_CTRL_FLEXEN_MASK);
+  ctrl_reg |= (FLEXIO_CTRL_DBGE(user_config->enable_indebug) |
+               FLEXIO_CTRL_FASTACC(user_config->enable_fast_access) |
+               FLEXIO_CTRL_FLEXEN(user_config->enable_flexio));
+  if (!user_config->enable_indoze)
+    {
+      ctrl_reg |= FLEXIO_CTRL_DOZEN_MASK;
+    }
+
+  base->CTRL = ctrl_reg;
+}
+
+/* Gets the default configuration to configure the FlexIO module.
+ *
+ * Example:
+ * code
+ * struct flexio_config_s config;
+ * imxrt_flexio_get_default_config(&config);
+ * endcode
+ *
+ * param user_config pointer to flexio_config_s structure
+ */
+
+static void imxrt_flexio_get_default_config(
+  struct flexio_config_s *user_config)
+{
+  assert(user_config != NULL);
+
+  (void)memset(user_config, 0, sizeof(*user_config));
```

Comment:
Why the cast is needed?

Reply:
> Why the cast is needed?

memset() returns the input pointer and the cast is probably to avoid unused return value warnings on some C/C++ compilers. But I think we have UNUSED() macro that is more suitable for that?

Reply:
I was just wondering why https://github.com/apache/nuttx/blob/0747ee1ea25f40336640eac47e61ec5aad65ec41/arch/arm/src/imxrt/imxrt_lcd.c#L248 from the same arch don't need a cast while it was suddenly needed here

Reply:
> I was just wondering why
> 
> https://github.com/apache/nuttx/blob/0747ee1ea25f40336640eac47e61ec5aad65ec41/arch/arm/src/imxrt/imxrt_ehci.c#L1080
> 
> from the same arch don't need a cast while it was suddenly needed here

Actually it is not needed here. Old habits:)